File: leetcode/rust-leetcode/src/328.odd-even-linked-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn odd_even_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.clone().is_none() || head.clone().unwrap().next.is_none() {
            return head;
        }
        let mut even = Box::new(ListNode::new(-1));
        let mut even_head = even.as_mut();
        let mut odd = Box::new(ListNode::new(-1));
        let mut odd_head = odd.as_mut();
        let mut even_flag = true;
        let mut dummy = head.unwrap();
        loop {
            let new_node = Some(Box::new(ListNode::new(dummy.val)));
            if even_flag {
                even_head.next = new_node;
                even_head = even_head.next.as_mut().unwrap().as_mut();
            } else {
                odd_head.next = new_node;
                odd_head = odd_head.next.as_mut().unwrap().as_mut();
            }
            even_flag = !even_flag;
            if dummy.next.is_none() {
                break;
            }
            dummy = dummy.next.unwrap();
        }
        even_head.next = odd.next;
        even.next
    }
}

File: leetcode/python/328.odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head
        odd = head
        even = head.next
        evenHead = even
        cur = head
        for _ in range(2):
            if cur is not None:
                cur = cur.next
        odd.next = None
        if even:
            even.next = None
        i = 1
        while cur is not None:
            if i == 1:
                odd.next = cur
                cur = cur.next
                odd = odd.next
                odd.next = None
                i = 0
            else:
                even.next = cur
                cur = cur.next
                even = even.next
                even.next = None
                i = 1
        odd.next = evenHead
        return head

